src/vmm/src/signal_handler.rs

```diff
@@ -128,162 +156,95 @@ generate_handler!(
+            libc::signalfd(-1, &mask as *const sigset_t, SFD_CLOEXEC | SFD_NONBLOCK)
+        })
+        .into_result()
+        .map_err(Error::SyscallError)?;
+
+        Ok(Self {
+            // Create a File instance so that we can leverage safe Read abstractions.
+            // Safe because the fd is valid.
+            signal_fd: unsafe { File::from_raw_fd(sfd) },
+        })
+    }
+}
 
-            // Call SIGXCPU signal handler.
-            assert_eq!(METRICS.signals.sigxcpu.count(), 0);
-            unsafe {
-                syscall(libc::SYS_kill, process::id(), SIGXCPU);
-            }
+impl Subscriber for SignalManager {
+    fn process(&mut self, event: &EpollEvent, _evmgr: &mut EventManager) {
+        let event_set = event.event_set();
+        let source_fd = event.fd();
 
-            // Call SIGPIPE signal handler.
-            assert_eq!(METRICS.signals.sigpipe.count(), 0);
-            unsafe {
-                syscall(libc::SYS_kill, process::id(), SIGPIPE);
-            }
+        if event_set != EventSet::IN || source_fd != self.signal_fd.as_raw_fd() {
+            error!("Spurious EventManager event for handler: SignalManager");
+            return;
+        }
 
-            // Call SIGHUP signal handler.
-            assert_eq!(METRICS.signals.sighup.count(), 0);
-            unsafe {
-                syscall(libc::SYS_kill, process::id(), SIGHUP);
-            }
+        let siginfo_size = mem::size_of::<signalfd_siginfo>();
+        let mut read_buf: Vec<u8> = vec![];
+        read_buf.resize(siginfo_size, 0);
 
-            // Call SIGILL signal handler.
-            assert_eq!(METRICS.signals.sigill.count(), 0);
-            unsafe {
-                syscall(libc::SYS_kill, process::id(), SIGILL);
-            }
-        });
-        assert!(child.join().is_ok());
-
-        // Sanity check.
-        assert!(cpu_count() > 0);
-        // Kcov somehow messes with our handler getting the SIGSYS signal when a bad syscall
-        // is caught, so the following assertion no longer holds. Ideally, we'd have a surefire
-        // way of either preventing this behaviour, or detecting for certain whether this test is
-        // run by kcov or not. The best we could do so far is to look at the perceived number of
-        // available CPUs. Kcov seems to make a single CPU available to the process running the
-        // tests, so we use this as an heuristic to decide if we check the assertion.
-        if cpu_count() > 1 {
-            // The signal handler should let the program continue during unit tests.
-            assert!(METRICS.seccomp.num_faults.count() >= 1);
+        // Even though we currently kill the process during signal handling, we should in theory
+        // read from the signalfd all the signals for future-proofing.
+        while self.signal_fd.read(&mut read_buf[..]).unwrap() == siginfo_size {
+            let siginfo: signalfd_siginfo = unsafe { ptr::read(read_buf.as_ptr() as *const _) };
```

Comment:
Is this safe ?

Reply:
In the latest version, I replaced this unsafe call with a call to the `ByteValued::from_slice()` which returns an option, so we are now good in terms of error handling. If the read is not possible, we should get a `None` and we log the error.